src/OmniSharp.Host/MSBuild/Discovery/Providers/VisualStudioInstanceProvider.cs

```diff
@@ -0,0 +1,89 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.Setup.Configuration;
+using OmniSharp.Utilities;
+
+namespace OmniSharp.MSBuild.Discovery.Providers
+{
+    internal class VisualStudioInstanceProvider : MSBuildInstanceProvider
+    {
+        public VisualStudioInstanceProvider(ILoggerFactory loggerFactory)
+            : base(loggerFactory)
+        {
+        }
+
+        public override ImmutableArray<MSBuildInstance> GetInstances()
+        {
+            if (!PlatformHelper.IsWindows)
+            {
+                return ImmutableArray<MSBuildInstance>.Empty;
+            }
+
+            try
+            {
+                var configuration = Interop.GetSetupConfiguration();
+                if (configuration == null)
+                {
+                    return ImmutableArray<MSBuildInstance>.Empty;
+                }
+
+                var builder = ImmutableArray.CreateBuilder<MSBuildInstance>();
+
+                var instanceEnum = configuration.EnumAllInstances();
+
+                int fetched;
+                var instances = new ISetupInstance[1];
+                do
+                {
+                    instanceEnum.Next(1, instances, out fetched);
```

Comment:
Does this only enumerate Dev15 versions?

Reply:
That's all it can do. This is the API for the Willow installer.

Reply:
Which gets picked if someone has dev15 and a dev15 preview installed?

Reply:
It's a bit arbitrary. We don't have a way for the user to select one, so it's whichever is first in the list. However, we check for the Visual Studio tied to a developer command prompt first. So, if the user wants to launch OmniSharp with a specific Visual Studio they can launch it from the command prompt that was installed for that VS. Alternatively, they can set the environment variables that the developer command prompt uses.